README.md
<h1 align="center">Welcome to mz-cli 👋</h1>
<p>
  <img alt="Version" src="https://img.shields.io/npm/v/mz-cli.svg">
  <img src="https://img.shields.io/badge/node-%3E%3D9.3-blue.svg" />
</p>

> cli tools

### 🏠 [Homepage](https://github.com/mwang0/mz-cli)

## Prerequisites

- node >=9.3

## Install

```sh
npm install mz-cli
```

## Author

👤 **M.Wang <cn_wang@139.com>**

* Github: [@mwang0](https://github.com/mwang0)

## 🤝 Contributing

Contributions, issues and feature requests are welcome!<br />Feel free to check [issues page](https://github.com/mwang0/mz-cli/issues).

## Show your support

Give a ⭐️ if this project helped you!

***
_This README was generated with ❤️ by [readme-md-generator](https://github.com/kefranabg/readme-md-generator)_

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.5.0](https://github.com/mwang0/mz-cli/compare/v0.4.0...v0.5.0) (2019-07-31)


### Bug Fixes

* 创建项目名为空出现异常 ([907feaf](https://github.com/mwang0/mz-cli/commit/907feaf)), closes [#1](https://github.com/mwang0/mz-cli/issues/1)


### Features

* 新增快应用工程模板 ([41b9ee9](https://github.com/mwang0/mz-cli/commit/41b9ee9))



## [0.4.0](https://github.com/mwang0/mz-cli/compare/v0.3.0...v0.4.0) (2019-07-24)


### Features

* **cli:** 增加:调用 meta.js 中 complete 逻辑 ([e548696](https://github.com/mwang0/mz-cli/commit/e548696))
* **utils.js:** 增加: isYarn\installDependencies\runCmd 方法 ([e33e101](https://github.com/mwang0/mz-cli/commit/e33e101))



## [0.3.0](https://github.com/mwang0/mz-cli/compare/v0.2.0...v0.3.0) (2019-07-23)


### Bug Fixes

* **spinner:** 终止 spinner 出现异常 ([e30036c](https://github.com/mwang0/mz-cli/commit/e30036c))


### Features

* **cli:** 重命名 mz -> mx ([f31ad93](https://github.com/mwang0/mz-cli/commit/f31ad93))
* **config:** 添加: standard-version 等依赖 ([70f4750](https://github.com/mwang0/mz-cli/commit/70f4750))
* **utils.js:** 增加对自已的更新检查 ([e9aec83](https://github.com/mwang0/mz-cli/commit/e9aec83))



## 0.2.0 (2019-07-21)


### Features

* **config:** 添加: list-staged ([d01d53d](https://github.com/mwang0/mz-cli/commit/d01d53d))
* **config:** 添加: prettier ([db2e6ba](https://github.com/mwang0/mz-cli/commit/db2e6ba))
* **mz-cli:** first commit ([60d9509](https://github.com/mwang0/mz-cli/commit/60d9509))


lib/create.js
/*
  create.js
  Created by M.Wang [cn_wang@139.com]
  2019-07-16 15:31 星期二
*/
const ejs = require('ejs')
const path = require('path')
const chalk = require('chalk')
const fg = require('fast-glob')
const fs = require('fs-extra')
const inquirer = require('inquirer')
const userHome = require('user-home')
const dl = require('download-git-repo')
const updateNotifier = require('update-notifier')
const validateProjectName = require('validate-npm-package-name')
const utils = require('./utils')
const logSpinner = require('./spinner')
const {
  exit,
  clearConsole,
  getTitle,
  getUserName,
  getUserEmail,
  trace,
  traceWarn,
  traceError,
  traceDone
} = utils
const { logWithSpinner, stopSpinner, changeSpinnerMsg } = logSpinner

async function create (projectName, options) {
  const cwd = process.cwd()
  const inCurrent = projectName === '.'
  const name = inCurrent ? path.relative('../', cwd) : projectName
  const targetDir = path.resolve(cwd, projectName || '.')

  // 项目名称校验
  const result = validateProjectName(name)
  if (!result.validForNewPackages) {
    traceError('无效的项目名称：' + name)
    result.errors &&
      result.errors.forEach((error) => {
        traceError(error)
      })
    result.warnings &&
      result.warnings.forEach((warn) => {
        traceWarn(warn)
      })
    exit(1)
  }

  if (fs.existsSync(targetDir)) {
    if (options.force) {
      await fs.remove(targetDir)
    } else {
      const title = await getTitle()
      await clearConsole(title)
      if (inCurrent) {
        const { ok } = await inquirer.prompt({
          name: 'ok',
          type: 'confirm',
          message: '在当前目录创建项目?'
        })

        if (!ok) {
          return
        }
      } else {
        const { action } = await inquirer.prompt({
          name: 'action',
          type: 'list',
          message: `项目目录 ${chalk.cyan(targetDir)} 已存在，请选择：`,
          choices: [
            { name: '覆盖', value: 'overwrite' },
            { name: '合并', value: 'merge' },
            { name: '取消', value: false }
          ]
        })

        if (!action) {
          return
        } else if (action === 'overwrite') {
          trace('♻️' + `  正在删除 ${chalk.cyan(targetDir)} ...`)
          await fs.remove(targetDir)
        }
      }
    }
  }

  await clearConsole(await getTitle())
  const { template } = await inquirer.prompt({
    type: 'list',
    name: 'template',
    message: '选择工程模板',
    choices: [
      { name: 'npm             ->  npm模块工程', value: 'npm' },
      { name: 'quickApp        ->  快应用工程', value: 'quick-app' },
      { name: 'template        ->  mz-cli模板工程', value: 'template' }
    ]
  })

  await clearConsole(await getTitle())
  logWithSpinner('✨', '项目创建在 ' + chalk.yellow(targetDir))
  logWithSpinner('💡', '项目类型为 ' + chalk.yellow(template))
  stopSpinner()
  trace()
  const tmpPath = path.join(userHome, '.m-templates', template)
  if (fs.existsSync(tmpPath)) {
    let pkg = path.join(tmpPath, 'package.json')
    if (fs.existsSync(pkg)) {
      pkg = require(pkg)
      const notifier = updateNotifier({
        pkg,
        updateCheckInterval: 1000 * 60 * 60 * 24
      })
      if (notifier.update) {
        trace(
          '🚀  ' +
            '更新项目【' +
            chalk.yellow(template) +
            '】模板：' +
            chalk.grey.dim(notifier.update.current) +
            ' ➟ ' +
            chalk.green.dim(notifier.update.latest)
        )
        await fs.remove(tmpPath)
        await download(template, tmpPath)
      }
    }
  } else {
    await download(template, tmpPath)
  }

  await generate(tmpPath, targetDir, { name })
  trace()
  traceDone('成功创建项目 ' + chalk.yellow(name))
  trace()
}

async function download (template, tmpsPath) {
  logWithSpinner('下载项目模板')
  await new Promise((resolve, reject) => {
    try {
      dl('m-templates/' + template, tmpsPath, null, (err) => {
        if (err) {
          stopSpinner({ symbol: '❌', text: '下载项目模板失败' })
          trace()
          traceError(err)
          reject(err)
          exit(1)
        }
        resolve()
      })
    } catch (error) {
      stopSpinner({ symbol: '❌', text: '下载项目模板失败' })
      trace()
      traceError(error)
      reject(error)
      exit(1)
    }
  })
  stopSpinner()
}

async function generate (tmpDir, targetDir, params = {}) {
  logWithSpinner('创建项目文件')
  const meta = path.join(tmpDir, 'meta.js')
  let opts = {}
  if (fs.existsSync(meta)) {
    const tmp = require(meta)
    if (tmp !== Object(tmp)) {
      trace()
      traceError('meta.js 需要导出一个对象')
      exit(1)
    }
    opts = tmp
  }
  const templateDir = path.join(tmpDir, 'template')
  if (!fs.existsSync(templateDir)) {
    trace()
    traceError('template目录不存在')
    exit(1)
  }
  params = { ...opts.preset, ...params }

  if (opts.prompts && Array.isArray(opts.prompts) && opts.prompts.length > 0) {
    params = await inquirer.prompt(opts)
  }
  params.username = await getUserName()
  params.userEmail = await getUserEmail()
  params.year = new Date().getFullYear()
  params.distDir = targetDir
  if (process.cwd() !== targetDir) {
    await fs.ensureDir(targetDir)
  }

  // 渲染模版文件
  const files = await fg(['**/*'], { cwd: templateDir, dot: true })
  await Promise.all(
    files.map((f) => {
      const fn = async () => {
        let content = await fs.readFile(path.join(templateDir, f), 'utf8')
        content = ejs.render(content, params)
        const filepath = path.join(targetDir, f)
        const fileDir = path.dirname(filepath)
        try {
          changeSpinnerMsg(f)
          await fs.ensureDir(fileDir)
          await fs.writeFile(filepath, content, { flag: 'a' })
        } catch (error) {
          trace()
          traceError(error)
        }
      }

      return fn()
    })
  )
  stopSpinner()
  if (opts.complete && typeof opts.complete === 'function') {
    try {
      await opts.complete(params, { ...utils, ...logSpinner, chalk })
    } catch (error) {
      traceError('运行 complete 异常...')
      traceError(error)
    }
  }
}

module.exports = (...args) => {
  return create(...args).catch((err) => {
    if (err) {
      trace()
      traceError(err)
    }
    process.exit(1)
  })
}


lib/spinner.js
const ora = require('ora')
const chalk = require('chalk')

const spinner = ora()
let lastMsg = null

exports.logWithSpinner = (symbol, msg) => {
  if (!msg) {
    msg = symbol
    symbol = chalk.green('✔')
  }
  if (lastMsg) {
    spinner.stopAndPersist({
      symbol: lastMsg.symbol,
      text: lastMsg.text
    })
  }
  spinner.text = ' ' + msg
  lastMsg = {
    symbol: symbol + ' ',
    text: msg
  }
  spinner.start()
}

exports.changeSpinnerMsg = (msg = '') => {
  spinner.text = lastMsg ? `${lastMsg}: ${msg}` : msg
}

exports.stopSpinner = (persist) => {
  if (lastMsg && persist !== false) {
    let { symbol, text } = lastMsg
    if (persist && 'symbol' in persist) symbol = persist.symbol
    if (persist && 'text' in persist) text = persist.text
    spinner.stopAndPersist({
      symbol,
      text
    })
  } else {
    spinner.stop()
  }
  lastMsg = null
}

exports.pauseSpinner = () => {
  spinner.stop()
}

exports.resumeSpinner = () => {
  spinner.start()
}


lib/utils.js
/*
  utils.js
  Created by M.Wang [cn_wang@139.com]
  2019-07-16 15:31 星期二
*/

const chalk = require('chalk')
const execa = require('execa')
const program = require('commander')
const readline = require('readline')
const spawn = require('child_process').spawn
const updateNotifier = require('update-notifier')

exports.exit = function (code = 0) {
  process.exit(code)
}

exports.clearConsole = (title) => {
  if (process.stdout.isTTY) {
    readline.cursorTo(process.stdout, 0, 0)
    readline.clearScreenDown(process.stdout)
    if (title) {
      console.log(title)
    }
    console.log()
  }
}

let versionCached
exports.getVersion = async function () {
  if (versionCached) return versionCached
  const pkg = require('../package.json')
  const local = pkg.version
  const notifier = updateNotifier({
    pkg,
    updateCheckInterval: 1000 * 60 * 60 * 24
  })
  versionCached = {
    current: local,
    latest: local
  }
  if (notifier.update) versionCached.latest = notifier.update.latest
  return versionCached
}

exports.getTitle = async function () {
  const { current, latest } = await exports.getVersion()
  let title = chalk.bold.blue('M CLI v' + current)
  if (latest !== current) {
    title += chalk.green(`
┌───────────${`─`.repeat(latest.length)}──┐
│  可用更新: ${latest}  │
└───────────${`─`.repeat(latest.length)}──┘`)
  }
  return title
}
// 获取 git user.name
exports.getUserName = async function () {
  const name = await exports.run('git', ['config', '--get', 'user.name'])
  return name || 'test'
}

// 获取 git user.email
exports.getUserEmail = async function () {
  const email = await exports.run('git', ['config', '--get', 'user.email'])
  return email || 'test@test.com'
}

// 执行 shell 命令
exports.run = async function (cmd, args, opts = {}) {
  if (!args) {
    [cmd, ...args] = cmd.split(/\s+/)
  }

  const { stdout } = await execa(cmd, args, opts)
  if (stdout) {
    return stdout
  }
  return null
}

exports.trace = (msg = '') => {
  console.log(msg)
}
exports.traceDone = (msg) => {
  console.log('🎉  ' + chalk.green(msg))
}
exports.traceWarn = (msg) => {
  console.log('⚠️  ', chalk.yellow(msg))
}
exports.traceError = (msg) => {
  console.log('❌  ', chalk.red(msg))
  if (msg instanceof Error) {
    console.error(msg.stack)
  }
}

exports.isYarn = async function isYarn () {
  let res = true
  try {
    await exports.run('yarn', ['-v'])
  } catch (err) {
    res = false
  }
  return res
}

/**
 * 运行指定命令
 * @date 2019-07-24
 * @param {string} cmd
 * @param {array<string>} args
 * @param {object} options
 * @returns {promise}
 */
exports.runCmd = function runCmd (cmd, args, options) {
  return new Promise((resolve, reject) => {
    const s = spawn(
      cmd,
      args,
      Object.assign(
        {
          cwd: process.cwd(),
          stdio: 'inherit',
          shell: true
        },
        options
      )
    )

    s.on('exit', () => {
      resolve()
    })
  })
}

/**
 * 依赖安装
 * @date 2019-07-23
 * @param {String} cwd
 * @returns {Promise}
 */
exports.installDependencies = async function installDependencies (cwd) {
  const { trace, runCmd, isYarn } = exports
  let executable = 'npm'
  let args = ['install']
  const isyarn = await isYarn()
  if (isyarn) {
    executable = 'yarn'
    args = []
  }
  trace(`\n'⭐️  安装项目依赖...'`)
  return runCmd(executable, args, { cwd })
}

exports.enhanceErrorMessages = function enhanceErrorMessages (methodName, log) {
  program.Command.prototype[methodName] = function (...args) {
    if (methodName === 'unknownOptions' && this._allowUnknowOptions) {
      // eslint-disable-next-line no-useless-return
      return
    }

    this.outputHelp()
    console.log(chalk.red(log(...args)))
    console.log()
    process.exit(1)
  }
}
